Cursor is not a Readable

### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

8.7.1

### Node.js version

20.18

### MongoDB server version

6.0.14

### Typescript version (if applicable)

5.6.3

### Description

When querying, using `.cursor()` used to return an `internal.Readable` compatible stream (which can be consumed in rxjs) but it seems not the case anymore.

By iterating on the type issues, here's what I've found to fix the issue:

Modify Cursor class definitions to be:

1. change: `[Symbol.asyncIterator](): Cursor<DocType, Options>;`
Since Cursor extends Readable, i think it's still valid

2. change: `next(): Promise<IteratorResult<Cursor<DocType, Options>, DocType>>;`
That was raised by TS after fixing the 1st.

### Steps to Reproduce

I'm still in the process of reproducing the case in an isolated environment. While waiting, here are the TS errors

Before fix number 1:
![Image](https://github.com/user-attachments/assets/fd784081-7dc1-495f-ac73-cf448837843f)

Before fix number 2:
![Image](https://github.com/user-attachments/assets/009f23cb-3c2d-4261-9389-f65a93e4d149)


### Expected Behavior

_No response_

I had also run into something similar to this, i think https://github.com/microsoft/TypeScript/issues/59263 is related, a workaround would be to switch your tsconfig `lib` from `esnext` to `es2023`.

for context, i got the following errors, even though i was not using any cursors directly:

```txt
node_modules/mongoose/types/cursor.d.ts:15:5 - error TS2416: Property '[Symbol.asyncIterator]' in type 'Cursor<DocType, Options>' is not assignable to the same property in base type 'Readable'.
  Type '() => AsyncIterableIterator<DocType>' is not assignable to type '() => AsyncIterator<any, any, any>'.
    Property '[Symbol.asyncDispose]' is missing in type 'AsyncIterableIterator<DocType>' but required in type 'AsyncIterator<any, any, any>'.

15     [Symbol.asyncIterator](): AsyncIterableIterator<DocType>;
       ~~~~~~~~~~~~~~~~~~~~~~

  node_modules/typescript/lib/lib.esnext.disposable.d.ts:40:5
    40     [Symbol.asyncDispose](): PromiseLike<void>;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.asyncDispose]' is declared here.

node_modules/mongoose/types/cursor.d.ts:55:5 - error TS2416: Property 'map' in type 'Cursor<DocType, Options>' is not assignable to the same property in base type 'Readable'.
  Type '<ResultType>(fn: (res: DocType) => ResultType) => Cursor<ResultType, Options>' is not assignable to type '(fn: (data: any, options?: Pick<ArrayOptions, "signal"> | undefined) => any, options?: ArrayOptions | undefined) => Readable'.
    Call signature return types 'Cursor<any, Options>' and 'Readable' are incompatible.
      The types returned by '[Symbol.asyncIterator]()' are incompatible between these types.
        Property '[Symbol.asyncDispose]' is missing in type 'AsyncIterableIterator<any>' but required in type 'AsyncIterator<any, any, any>'.

55     map<ResultType>(fn: (res: DocType) => ResultType): Cursor<ResultType, Options>;
       ~~~

  node_modules/typescript/lib/lib.esnext.disposable.d.ts:40:5
    40     [Symbol.asyncDispose](): PromiseLike<void>;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.asyncDispose]' is declared here.
```
(`typescript@5.6.3` and `@types/node@22.8.7`)

I am also facing probably the same issue. My environments is:

Node `v20.18.0`, Typescript: `v5.4.5`, `@types/node: v20.14.8`, Mongoose: `v8.4.3`. Below is my stacktrace for build operation.

```plaintext
node_modules/mongoose/types/cursor.d.ts:14:5 - error TS2416: Property '[Symbol.asyncIterator]' in type 'Cursor<DocType, Options>' is not assignable to the same property in base type 'Readable'.
  Type '() => AsyncIterableIterator<DocType>' is not assignable to type '() => AsyncIterator<any, any, any>'.
    Property '[Symbol.asyncDispose]' is missing in type 'AsyncIterableIterator<DocType>' but required in type 'AsyncIterator<any, any, any>'.

14     [Symbol.asyncIterator](): AsyncIterableIterator<DocType>;
       ~~~~~~~~~~~~~~~~~~~~~~

  node_modules/typescript/lib/lib.esnext.disposable.d.ts:40:5
    40     [Symbol.asyncDispose](): PromiseLike<void>;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.asyncDispose]' is declared here.

node_modules/mongoose/types/cursor.d.ts:48:5 - error TS2416: Property 'map' in type 'Cursor<DocType, Options>' is not assignable to the same property in base type 'Readable'.
  Type '<ResultType>(fn: (res: DocType) => ResultType) => Cursor<ResultType, Options>' is not assignable to type '(fn: (data: any, options?: Pick<ArrayOptions, "signal"> | undefined) => any, options?: ArrayOptions | undefined) => Readable'.
    Call signature return types 'Cursor<any, Options>' and 'Readable' are incompatible.
      The types returned by '[Symbol.asyncIterator]()' are incompatible between these types.
        Property '[Symbol.asyncDispose]' is missing in type 'AsyncIterableIterator<any>' but required in type 'AsyncIterator<any, any, any>'.

48     map<ResultType>(fn: (res: DocType) => ResultType): Cursor<ResultType, Options>;
       ~~~

  node_modules/typescript/lib/lib.esnext.disposable.d.ts:40:5
    40     [Symbol.asyncDispose](): PromiseLike<void>;
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.asyncDispose]' is declared here.


Found 2 errors in the same file, starting at: node_modules/mongoose/types/cursor.d.ts:14
```

--- a/types/cursor.d.ts
+++ b/types/cursor.d.ts
@@ -11,8 +11,10 @@ declare module 'mongoose' {
     signal?: AbortSignal;
   }
 
-  class Cursor<DocType = any, Options = never> extends stream.Readable {
-    [Symbol.asyncIterator](): AsyncIterableIterator<DocType>;
+  class Cursor<DocType = any, Options = never, NextResultType = DocType | null> extends stream.Readable {
+    [Symbol.asyncIterator](): Cursor<IteratorResult<DocType>, Options, IteratorResult<DocType>>;
+
+    [Symbol.asyncDispose](): Promise<void>;
 
     /**
      * Adds a [cursor flag](https://mongodb.github.io/node-mongodb-native/4.9/classes/FindCursor.html#addCursorFlag).
@@ -58,7 +60,7 @@ declare module 'mongoose' {
      * Get the next document from this cursor. Will return `null` when there are
      * no documents left.
      */
-    next(): Promise<DocType | null>;
+    next(): Promise<NextResultType>;
 
     options: Options;
   }

